src/GraphCoarsen.cpp

```diff
@@ -94,6 +98,25 @@ void GraphCoarsen::BuildPVertices(
             {
                 Pvertices_j[ptr] = coarse_vertex_dof_counter + k;
                 Pvertices_data[ptr++] = target_i(j, k);
+                if (k == 0)
+                {
+                    const double cval = coarse_constant_rep_(coarse_vertex_dof_counter);
+                    if (cval == 0.0)
+                    {
+                        // @todo this should probably depend on volume size etc?
```

Comment:
It may not depend on the volume of elements in fine level (unless the basis for the fine level is not constant 1 on an element), but it will depend on the volume of aggregate, which I guess is why you take `1.0 / target_i(j, k)` here.

An alternative way to compute the representation of constant vector in coarse level is to make use the fact that `Pvertices` is orthogonal, so

constant_one_coarse = Pvertices * constant_one_fine 

Then the volume factor will be taken care of automatically, and it should be extensible to multilevel as long as we construct `Pvertices` properly.

Reply:
I feel fairly dumb for not realizing this myself.